TYPO3-Solr/ext-solr
# [FEATURE] use assoc keys for tx_solr[filter] facet parameters instead of numeric index

Introduce a new style how the facet array represented inside of the url with a combination of key and value in order to be able to keep a specific order.

# What this pr does

This pull request allows to change the url parameters from index based to associative keys for facets.
Using associative keys, the value of a facet will be replaced with 1. A value of 1 means, that the facet is active. A value of 0 means, the value is inactive.

Additionaly a new sort option for url parameters is available. The sort of parameters is mandatory for associative keys.

By default Solr behaves as before.

**Note:** This change affects the representation of facets inside of the URL strings.

# How to test

1. Include all possibile facet configuration in your SysTemplate: 

```
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/DateRange/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/Hierarchy/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/NumericRange/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/Options/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/OptionsFiltered/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/OptionsPrefixGrouped/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/OptionsSinglemode/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/OptionsToggle/setup.txt'
@import 'EXT:solr/Configuration/TypoScript/Examples/Facets/QueryGroup/setup.txt'
```
2. Index all pages and search for `*`
3. Select facets in following order
  1. Type
  2. Age
  3. PID Range

As a result the order of filter inside of the URL have to be:
  1. Age
  2. PID Range
  3. Type

* Now click on page number two. The order of parameters should not change.
* Remove facet PID Range.
As a result the order of filter inside of the URL have to be:
  1. Age
  2. Type

* Run all unit test. You should not see any errors.
* Run all integration test. You should not see any errors.


Fixes: #2307
